tests/pytests/unit/fileserver/test_s3fs.py

```diff
@@ -1,19 +1,154 @@
+        conn = boto3.client("s3", **aws_creds)
+        conn.create_bucket(Bucket=bucket)
+        return conn
+
+    return _s3
+
+
+def make_keys(bucket, conn, keys):
+    for key, data in keys.items():
+        conn.put_object(
+            Bucket=bucket,
+            Key=key,
+            Body=data["content"],
+        )
+
+
+def verify_cache(bucket, expected):
+    for key, data in expected.items():
+        correct_content = data["content"]
+        cache_file = s3fs._get_cached_file_name(bucket, "base", key)
+        assert os.path.exists(cache_file)
+
+        if correct_content is None:
+            continue
+
+        with salt.utils.files.fopen(cache_file) as f:
+            content = f.read()
+            assert correct_content == content
+
+
+# [HACK] to always invalidate the cache and re-download on every update
+@pytest.fixture(scope="module", autouse=True)
+def _patch_cache_expire():
+    with patch("salt.fileserver.s3fs.S3_CACHE_EXPIRE", -1):
+        yield
+
+
+@mock_s3
+def test_update(bucket, s3):
+    """Tests that files get downloaded from s3 to the local cache."""
+
+    keys = {
+        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
+        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
+        "files/nginx.conf": {"content": "server {}"},
+    }
+
+    conn = s3()
+    make_keys(bucket, conn, keys)
+    s3fs.update()
+    verify_cache(bucket, keys)
+
+    # make a modification and update again - verify the change is retrieved
+    keys["top.sls"]["content"] = yaml.dump({"base": {"*": ["foo", "bar"]}})
+    make_keys(bucket, conn, keys)
+    s3fs.update()
+    verify_cache(bucket, keys)
+
+    # TODO: fix
+
+    # delete_file = "files/nginx.conf"
+    # del keys[delete_file]
+    # conn.delete_object(Bucket=bucket, Key=delete_file)
+
+    # s3fs.update()
+    # verify_cache(bucket, keys)
+
+    # cache_file = s3fs._get_cached_file_name(bucket, "base", delete_file)
+    # assert not os.path.exists(cache_file)
+
+
+@mock_s3
+def test_s3_hash(bucket, s3):
+    """Verifies that s3fs hashes files correctly."""
+
+    keys = {
+        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
+        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
+        "files/nginx.conf": {"content": "server {}"},
+    }
+
+    conn = s3()
+    make_keys(bucket, conn, keys)
+    s3fs.update()
+
+    for key, item in keys.items():
+        cached_file_path = s3fs._get_cached_file_name(bucket, "base", key)
+        item["hash"] = salt.utils.hashutils.get_hash(
+            cached_file_path, s3fs.S3_HASH_TYPE
+        )
+        item["cached_file_path"] = cached_file_path
+
+    load = {"saltenv": "base"}
+    fnd = {"bucket": bucket}
+
+    for key, item in keys.items():
+        fnd["path"] = item["cached_file_path"]
+        actual_hash = s3fs.file_hash(load, fnd)
+        assert s3fs.S3_HASH_TYPE == actual_hash["hash_type"]
+        assert item["hash"] == actual_hash["hsum"]
```

Comment:
```suggestion
@pytest.fixture
def _mock_s3():
    with mock_s3():
        yield


@pytest.mark.usefixture("_mock_s3")
def test_update(bucket, s3):
    """Tests that files get downloaded from s3 to the local cache."""

    keys = {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }

    conn = s3()
    make_keys(bucket, conn, keys)
    s3fs.update()
    verify_cache(bucket, keys)

    # make a modification and update again - verify the change is retrieved
    keys["top.sls"]["content"] = yaml.dump({"base": {"*": ["foo", "bar"]}})
    make_keys(bucket, conn, keys)
    s3fs.update()
    verify_cache(bucket, keys)

    # TODO: fix

    # delete_file = "files/nginx.conf"
    # del keys[delete_file]
    # conn.delete_object(Bucket=bucket, Key=delete_file)

    # s3fs.update()
    # verify_cache(bucket, keys)

    # cache_file = s3fs._get_cached_file_name(bucket, "base", delete_file)
    # assert not os.path.exists(cache_file)


@pytest.mark.usefixture("_mock_s3")
def test_s3_hash(bucket, s3):
    """Verifies that s3fs hashes files correctly."""

    keys = {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }

    conn = s3()
    make_keys(bucket, conn, keys)
    s3fs.update()

    for key, item in keys.items():
        cached_file_path = s3fs._get_cached_file_name(bucket, "base", key)
        item["hash"] = salt.utils.hashutils.get_hash(
            cached_file_path, s3fs.S3_HASH_TYPE
        )
        item["cached_file_path"] = cached_file_path

    load = {"saltenv": "base"}
    fnd = {"bucket": bucket}

    for key, item in keys.items():
        fnd["path"] = item["cached_file_path"]
        actual_hash = s3fs.file_hash(load, fnd)
        assert s3fs.S3_HASH_TYPE == actual_hash["hash_type"]
        assert item["hash"] == actual_hash["hsum"]
```

http://docs.getmoto.org/en/latest/docs/getting_started.html?highlight=mock_s3#context-manager

Or, alternatively(to not have to use `@pytest.mark.usefixture("_mock_s3")`:
```python
@pytest.fixture
def bucket():
    with mock_s3():
        yield "bucket-name"
```

Reply:
Done